File: typespec/hub/incognito.go
package hub

import (
	"time"

	"github.com/vetchium/vetchium/typespec/common"
)

type IncognitoPost struct {
	IncognitoPostID string        `json:"incognito_post_id"`
	Content         string        `json:"content"`
	Tags            []common.VTag `json:"tags"`
	CreatedAt       time.Time     `json:"created_at"`
	UpvotesCount    int32         `json:"upvotes_count"`
	DownvotesCount  int32         `json:"downvotes_count"`
	Score           int32         `json:"score"`
	MeUpvoted       bool          `json:"me_upvoted"`
	MeDownvoted     bool          `json:"me_downvoted"`
	CanUpvote       bool          `json:"can_upvote"`
	CanDownvote     bool          `json:"can_downvote"`
	IsCreatedByMe   bool          `json:"is_created_by_me"`
	IsDeleted       bool          `json:"is_deleted"`
}

type AddIncognitoPostRequest struct {
	Content string          `json:"content" validate:"required,min=1,max=1024"`
	TagIDs  []common.VTagID `json:"tag_ids" validate:"min=1,max=3"`
}

type AddIncognitoPostResponse struct {
	IncognitoPostID string `json:"incognito_post_id"`
}

type IncognitoPostComment struct {
	CommentID      string    `json:"comment_id"`
	Content        string    `json:"content"`
	InReplyTo      *string   `json:"in_reply_to"`
	CreatedAt      time.Time `json:"created_at"`
	UpvotesCount   int32     `json:"upvotes_count"`
	DownvotesCount int32     `json:"downvotes_count"`
	Score          int32     `json:"score"`
	MeUpvoted      bool      `json:"me_upvoted"`
	MeDownvoted    bool      `json:"me_downvoted"`
	CanUpvote      bool      `json:"can_upvote"`
	CanDownvote    bool      `json:"can_downvote"`
	IsCreatedByMe  bool      `json:"is_created_by_me"`
	IsDeleted      bool      `json:"is_deleted"`
	Depth          int32     `json:"depth"`
	RepliesCount   int32     `json:"replies_count"`
}

type AddIncognitoPostCommentRequest struct {
	IncognitoPostID string  `json:"incognito_post_id" validate:"required"`
	Content         string  `json:"content"           validate:"required,min=1,max=512"`
	InReplyTo       *string `json:"in_reply_to"`
}

type AddIncognitoPostCommentResponse struct {
	IncognitoPostID string `json:"incognito_post_id"`
	CommentID       string `json:"comment_id"`
}

type IncognitoPostCommentSortBy string

const (
	IncognitoPostCommentSortByTop IncognitoPostCommentSortBy = "top"
	IncognitoPostCommentSortByNew IncognitoPostCommentSortBy = "new"
	IncognitoPostCommentSortByOld IncognitoPostCommentSortBy = "old"
)

type GetIncognitoPostCommentsRequest struct {
	IncognitoPostID         string                     `json:"incognito_post_id"          validate:"required"`
	SortBy                  IncognitoPostCommentSortBy `json:"sort_by"`
	PaginationKey           *string                    `json:"pagination_key"`
	Limit                   int32                      `json:"limit"                      validate:"min=1,max=50"`
	DirectRepliesPerComment int32                      `json:"direct_replies_per_comment" validate:"min=0,max=10"`
}

type GetIncognitoPostCommentsResponse struct {
	Comments           []IncognitoPostComment `json:"comments"`
	PaginationKey      string                 `json:"pagination_key"`
	TotalCommentsCount int32                  `json:"total_comments_count"`
}

type GetCommentRepliesRequest struct {
	IncognitoPostID string  `json:"incognito_post_id" validate:"required"`
	ParentCommentID string  `json:"parent_comment_id" validate:"required"`
	PaginationKey   *string `json:"pagination_key"`
	Limit           int32   `json:"limit"             validate:"min=1,max=100"`
	DirectOnly      bool    `json:"direct_only"`
	MaxDepth        int32   `json:"max_depth"         validate:"min=1,max=3"`
}

type GetCommentRepliesResponse struct {
	Replies           []IncognitoPostComment `json:"replies"`
	PaginationKey     string                 `json:"pagination_key"`
	TotalRepliesCount int32                  `json:"total_replies_count"`
	ParentCommentID   string                 `json:"parent_comment_id"`
}

type GetIncognitoPostCommentPermalinkRequest struct {
	IncognitoPostID      string `json:"incognito_post_id"      validate:"required"`
	CommentID            string `json:"comment_id"             validate:"required"`
	ContextSiblingsCount int32  `json:"context_siblings_count" validate:"min=0,max=10"`
	ContextRepliesCount  int32  `json:"context_replies_count"  validate:"min=0,max=20"`
}

type GetIncognitoPostCommentPermalinkResponse struct {
	Comments        []IncognitoPostComment `json:"comments"`
	TargetCommentID string                 `json:"target_comment_id"`
	BreadcrumbPath  []string               `json:"breadcrumb_path"`
}

type DeleteIncognitoPostCommentRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
	CommentID       string `json:"comment_id"        validate:"required"`
}

type UpvoteIncognitoPostCommentRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
	CommentID       string `json:"comment_id"        validate:"required"`
}

type DownvoteIncognitoPostCommentRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
	CommentID       string `json:"comment_id"        validate:"required"`
}

type UnvoteIncognitoPostCommentRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
	CommentID       string `json:"comment_id"        validate:"required"`
}

type GetIncognitoPostRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
}

type DeleteIncognitoPostRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
}

type UpvoteIncognitoPostRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
}

type DownvoteIncognitoPostRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
}

type UnvoteIncognitoPostRequest struct {
	IncognitoPostID string `json:"incognito_post_id" validate:"required"`
}

type MyIncognitoPostComment struct {
	CommentID          string        `json:"comment_id"`
	Content            string        `json:"content"`
	InReplyTo          *string       `json:"in_reply_to"`
	CreatedAt          time.Time     `json:"created_at"`
	UpvotesCount       int32         `json:"upvotes_count"`
	DownvotesCount     int32         `json:"downvotes_count"`
	Score              int32         `json:"score"`
	MeUpvoted          bool          `json:"me_upvoted"`
	MeDownvoted        bool          `json:"me_downvoted"`
	IsDeleted          bool          `json:"is_deleted"`
	Depth              int32         `json:"depth"`
	IncognitoPostID    string        `json:"incognito_post_id"`
	PostContentPreview string        `json:"post_content_preview"`
	PostTags           []common.VTag `json:"post_tags"`
}

type GetMyIncognitoPostCommentsRequest struct {
	PaginationKey *string `json:"pagination_key"`
	Limit         int32   `json:"limit"          validate:"min=1,max=40"`
}

type GetMyIncognitoPostCommentsResponse struct {
	Comments      []MyIncognitoPostComment `json:"comments"`
	PaginationKey string                   `json:"pagination_key"`
}

type IncognitoPostTimeFilter string

const (
	IncognitoPostTimeFilterPast24Hours IncognitoPostTimeFilter = "past_24_hours"
	IncognitoPostTimeFilterPastWeek    IncognitoPostTimeFilter = "past_week"
	IncognitoPostTimeFilterPastMonth   IncognitoPostTimeFilter = "past_month"
	IncognitoPostTimeFilterPastYear    IncognitoPostTimeFilter = "past_year"
)

type GetIncognitoPostsRequest struct {
	TagID         common.VTagID            `json:"tag_id"         validate:"required"`
	TimeFilter    *IncognitoPostTimeFilter `json:"time_filter"`
	Limit         int32                    `json:"limit"          validate:"min=1,max=100"`
	PaginationKey *string                  `json:"pagination_key"`
}

type IncognitoPostSummary struct {
	IncognitoPostID string        `json:"incognito_post_id"`
	Content         string        `json:"content"`
	Tags            []common.VTag `json:"tags"`
	CreatedAt       time.Time     `json:"created_at"`
	UpvotesCount    int32         `json:"upvotes_count"`
	DownvotesCount  int32         `json:"downvotes_count"`
	Score           int32         `json:"score"`
	MeUpvoted       bool          `json:"me_upvoted"`
	MeDownvoted     bool          `json:"me_downvoted"`
	CanUpvote       bool          `json:"can_upvote"`
	CanDownvote     bool          `json:"can_downvote"`
	CommentsCount   int32         `json:"comments_count"`
	IsCreatedByMe   bool          `json:"is_created_by_me"`
	IsDeleted       bool          `json:"is_deleted"`
}

type GetIncognitoPostsResponse struct {
	Posts         []IncognitoPostSummary `json:"posts"`
	PaginationKey string                 `json:"pagination_key"`
}

type GetMyIncognitoPostsRequest struct {
	PaginationKey *string `json:"pagination_key"`
	Limit         int32   `json:"limit"          validate:"min=1,max=40"`
}

type GetMyIncognitoPostsResponse struct {
	Posts         []IncognitoPostSummary `json:"posts"`
	PaginationKey string                 `json:"pagination_key"`
}

File: typespec/hub/incognito.ts
import { VTag, VTagID } from "../common/vtags";

export class IncognitoPost {
  incognito_post_id: string = "";
  content: string = "";
  tags: VTag[] = [];
  created_at: string = "";
  upvotes_count: number = 0;
  downvotes_count: number = 0;
  score: number = 0;
  me_upvoted: boolean = false;
  me_downvoted: boolean = false;
  can_upvote: boolean = false;
  can_downvote: boolean = false;
  is_created_by_me: boolean = false;
  is_deleted: boolean = false;
}

export class AddIncognitoPostRequest {
  content: string = "";
  tag_ids: VTagID[] = [];

  IsValid(): boolean {
    return (
      this.content.length > 0 &&
      this.content.length <= 1024 &&
      this.tag_ids.length >= 1 &&
      this.tag_ids.length <= 3
    );
  }
}

export class AddIncognitoPostResponse {
  incognito_post_id: string = "";
}

export class IncognitoPostComment {
  comment_id: string = "";
  content: string = "";
  in_reply_to?: string = undefined;
  created_at: string = "";
  upvotes_count: number = 0;
  downvotes_count: number = 0;
  score: number = 0;
  me_upvoted: boolean = false;
  me_downvoted: boolean = false;
  can_upvote: boolean = false;
  can_downvote: boolean = false;
  is_created_by_me: boolean = false;
  is_deleted: boolean = false;
  depth: number = 0;
  replies_count: number = 0;
}

export class AddIncognitoPostCommentRequest {
  incognito_post_id: string = "";
  content: string = "";
  in_reply_to?: string = undefined;

  IsValid(): boolean {
    return (
      this.incognito_post_id.length > 0 &&
      this.content.length > 0 &&
      this.content.length <= 512
    );
  }
}

export class AddIncognitoPostCommentResponse {
  incognito_post_id: string = "";
  comment_id: string = "";
}

export enum IncognitoPostCommentSortBy {
  Top = "top",
  New = "new",
  Old = "old",
}

export class GetIncognitoPostCommentsRequest {
  incognito_post_id: string = "";
  sort_by: IncognitoPostCommentSortBy = IncognitoPostCommentSortBy.Top;
  pagination_key?: string = undefined;
  limit: number = 25;
  direct_replies_per_comment: number = 3;

  IsValid(): boolean {
    return (
      this.incognito_post_id.length > 0 &&
      this.limit >= 1 &&
      this.limit <= 50 &&
      this.direct_replies_per_comment >= 0 &&
      this.direct_replies_per_comment <= 10
    );
  }
}

export class GetIncognitoPostCommentsResponse {
  comments: IncognitoPostComment[] = [];
  pagination_key: string = "";
  total_comments_count: number = 0;
}

export class GetCommentRepliesRequest {
  incognito_post_id: string = "";
  parent_comment_id: string = "";
  pagination_key?: string = undefined;
  limit: number = 50;
  direct_only: boolean = true;
  max_depth: number = 2;

  IsValid(): boolean {
    return (
      this.incognito_post_id.length > 0 &&
      this.parent_comment_id.length > 0 &&
      this.limit >= 1 &&
      this.limit <= 100 &&
      this.max_depth >= 1 &&
      this.max_depth <= 3
    );
  }
}

export class GetCommentRepliesResponse {
  replies: IncognitoPostComment[] = [];
  pagination_key: string = "";
  total_replies_count: number = 0;
  parent_comment_id: string = "";
}

export class GetIncognitoPostCommentPermalinkRequest {
  incognito_post_id: string = "";
  comment_id: string = "";
  context_siblings_count: number = 3;
  context_replies_count: number = 10;

  IsValid(): boolean {
    return (
      this.incognito_post_id.length > 0 &&
      this.comment_id.length > 0 &&
      this.context_siblings_count >= 0 &&
      this.context_siblings_count <= 10 &&
      this.context_replies_count >= 0 &&
      this.context_replies_count <= 20
    );
  }
}

export class GetIncognitoPostCommentPermalinkResponse {
  comments: IncognitoPostComment[] = [];
  target_comment_id: string = "";
  breadcrumb_path: string[] = [];
}

export class DeleteIncognitoPostCommentRequest {
  incognito_post_id: string = "";
  comment_id: string = "";
}

export class UpvoteIncognitoPostCommentRequest {
  incognito_post_id: string = "";
  comment_id: string = "";
}

export class DownvoteIncognitoPostCommentRequest {
  incognito_post_id: string = "";
  comment_id: string = "";
}

export class UnvoteIncognitoPostCommentRequest {
  incognito_post_id: string = "";
  comment_id: string = "";
}

export class GetIncognitoPostRequest {
  incognito_post_id: string = "";
}

export class DeleteIncognitoPostRequest {
  incognito_post_id: string = "";
}

export class UpvoteIncognitoPostRequest {
  incognito_post_id: string = "";
}

export class DownvoteIncognitoPostRequest {
  incognito_post_id: string = "";
}

export class UnvoteIncognitoPostRequest {
  incognito_post_id: string = "";
}

export class MyIncognitoPostComment {
  comment_id: string = "";
  content: string = "";
  in_reply_to?: string = undefined;
  created_at: string = "";
  upvotes_count: number = 0;
  downvotes_count: number = 0;
  score: number = 0;
  me_upvoted: boolean = false;
  me_downvoted: boolean = false;
  is_deleted: boolean = false;
  depth: number = 0;
  incognito_post_id: string = "";
  post_content_preview: string = "";
  post_tags: VTag[] = [];
}

export class GetMyIncognitoPostCommentsRequest {
  pagination_key?: string = undefined;
  limit: number = 25;

  IsValid(): boolean {
    return this.limit >= 1 && this.limit <= 40;
  }
}

export class GetMyIncognitoPostCommentsResponse {
  comments: MyIncognitoPostComment[] = [];
  pagination_key: string = "";
}

export enum IncognitoPostTimeFilter {
  Past24Hours = "past_24_hours",
  PastWeek = "past_week",
  PastMonth = "past_month",
  PastYear = "past_year",
}

export class GetIncognitoPostsRequest {
  tag_id: VTagID = "";
  time_filter?: IncognitoPostTimeFilter = IncognitoPostTimeFilter.Past24Hours;
  limit: number = 25;
  pagination_key?: string = undefined;

  IsValid(): boolean {
    return this.tag_id.length > 0 && this.limit >= 1 && this.limit <= 100;
  }
}

export class IncognitoPostSummary {
  incognito_post_id: string = "";
  content: string = "";
  tags: VTag[] = [];
  created_at: string = "";
  upvotes_count: number = 0;
  downvotes_count: number = 0;
  score: number = 0;
  me_upvoted: boolean = false;
  me_downvoted: boolean = false;
  can_upvote: boolean = false;
  can_downvote: boolean = false;
  comments_count: number = 0;
  is_created_by_me: boolean = false;
  is_deleted: boolean = false;
}

export class GetIncognitoPostsResponse {
  posts: IncognitoPostSummary[] = [];
  pagination_key: string = "";
}

export class GetMyIncognitoPostsRequest {
  pagination_key?: string = undefined;
  limit: number = 25;

  IsValid(): boolean {
    return this.limit >= 1 && this.limit <= 40;
  }
}

export class GetMyIncognitoPostsResponse {
  posts: IncognitoPostSummary[] = [];
  pagination_key: string = "";
}
